contracts/Migrator.sol

```diff
@@ -0,0 +1,226 @@
+    event Decommissioned();
+    event LockupsMigrated(address indexed user, uint256[] ogvLockupIds, uint256 newDuration);
+
+    error MigrationAlreadyStarted();
+    error MigrationIsInactive();
+    error MigrationNotComplete();
+    error BalanceMismatch(uint256 ogvAmountIn, uint256 expectedOgnAmountOut, uint256 actualOgnAmountOut);
+    error LockupIdsRequired();
+
+    constructor(address _ogv, address _ogn, address _ogvStaking, address _ognStaking) {
+        ogv = ERC20Burnable(_ogv);
+        ogn = ERC20Burnable(_ogn);
+        ogvStaking = IStaking(_ogvStaking);
+        ognStaking = IStaking(_ognStaking);
+    }
+
+    /**
+     * @notice Starts the migration and sets it to end after
+     *          365 days. Also, approves xOGN to transfer OGN
+     *          held in this contract. Can be invoked only once
+     */
+    function start() external onlyGovernor {
+        if (endTime != 0) {
+            revert MigrationAlreadyStarted();
+        }
+
+        // Max approve
+        ogn.approve(address(ognStaking), type(uint256).max);
+
+        endTime = block.timestamp + 365 days;
+    }
+
+    /**
+     * @notice Decommissions the contract. Can be called only
+     *          after a year since `start()` was invoked. Burns
+     *          all OGV held in the contract and transfers OGN
+     *          to address(1).
+     */
+    function decommission() external {
+        // Only after a year of staking
+        if (endTime == 0 || isMigrationActive()) {
+            revert MigrationNotComplete();
+        }
+
+        emit Decommissioned();
+
+        uint256 ogvBalance = ogv.balanceOf(address(this));
+        if (ogvBalance > 0) {
+            // Burn all OGV
+            ogv.burn(ogvBalance);
+        }
+
+        uint256 ognBalance = ogn.balanceOf(address(this));
+        if (ognBalance > 0) {
+            // OGN doesn't allow burning of tokens. Has `onlyOwner`
+            // modifier on `burn` and `burnFrom` methods. Also,
+            // `transfer` has a address(0) check. So, this transfers
+            // everything to address(1). The `owner` multisig of
+            // OGN token can call `burnFrom(address(1))` later.abi
+
+            ogn.transfer(address(1), ognBalance);
+        }
+    }
+
+    /**
+     * @notice Returns the active status of the migration.
+     * @return True if migration has started and has not ended yet.
+     */
+    function isMigrationActive() public view returns (bool) {
+        return endTime > 0 && block.timestamp < endTime;
+    }
+
+    /**
+     * @notice Solvency Checks
+     *
+     * This ensures that the contract never transfers more than
+     * desired OGN amount in any case. This takes a balance diff
+     * of OGV and OGN and makes sure that the difference adds up.
+     * However, it doesn't revert if the difference is in favour
+     * of the contract (i.e. less OGN sent than expected).
+     */
+    modifier netBalanceCheck() {
```

Comment:
Let's move a modifier to the front of the contract. As we discussed, let's make this an isSolvent check that ensures that all future migration can happen.

Reply:
Have made that change. Now the contract only checks if it can fund the migration of OGV still in circulation (or staked)

Reply:
Checking against totalSupply is beautiful way to do this. Way better than what I was thinking.